File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.cs
using System;
class MainApp
{
    static int Blue;
    static int White;
    static int[,] Map;
    static void Cut(int A, int B, int N)
    {
        int Pivot = Map[A, B];
        bool Check = true;
        for (int i = A; i < A + N; i++)
        {
            for (int j = B; j < B + N; j++)
            {
                if (Map[i, j] != Pivot)
                    Check = false;
            }
        }
        if (Check)
        {
            if (Pivot == 1) Blue++;
            else White++;
        }
        else
        {
            Cut(A, B, N / 2);
            Cut(A + N / 2, B, N / 2);
            Cut(A, B + N / 2, N / 2);
            Cut(A + N / 2, B + N / 2, N / 2);
        }
    }
    static void Main(string[] args)
    {
        int N = int.Parse(Console.ReadLine());
        Map = new int[N, N];
        for (int i = 0; i < N; i++)
        {
            string[] Input = Console.ReadLine().Split();
            for (int j = 0; j < N; j++)
            {
                Map[i, j] = int.Parse(Input[j]);
            }
        }
        Cut(0, 0, N);
        Console.WriteLine(White);
        Console.WriteLine(Blue);
    }
}

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.io.*;
import java.util.*;

public class Main {
    static int white, blue;
    static int[][] map;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());
        map = new int[N][N];
        for (int i = 0; i < N; ++i) {
            st = new StringTokenizer(br.readLine(), " ");
            for (int j = 0; j < N; ++j) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        cntPaper(0, 0, N);
        sb.append(white).append("\n").append(blue);
        System.out.println(sb);
    }

    static void cntPaper(int x, int y, int k) {
        int pivot = map[y][x];
        boolean check = true;
        a:
        for (int i = x; i < x + k; ++i) {
            for (int j = y; j < y + k; ++j) {
                if (pivot != map[j][i]) {
                    check = false;
                    break a;
                }
            }
        }
        if (check) {
            if (pivot == 1) blue++;
            else white++;
        }
        else {
            cntPaper(x, y, k / 2);
            cntPaper(x, y + k / 2, k / 2);
            cntPaper(x + k / 2, y, k / 2);
            cntPaper(x + k / 2, y + k / 2, k / 2);
        }
    }
}